src/core/pointcloud/qgspointcloudlayerrenderer.cpp

```diff
@@ -140,7 +143,62 @@ bool QgsPointCloudLayerRenderer::render()
     mAttributes.push_back( mLayerAttributes.at( layerIndex ) );
   }
 
-  QgsPointCloudDataBounds db;
+  QgsRectangle renderExtent;
+  try
+  {
+    renderExtent = renderContext()->coordinateTransform().transformBoundingBox( renderContext()->mapExtent(), Qgis::TransformDirection::Reverse );
+  }
+  catch ( QgsCsException & )
+  {
+    QgsDebugMsg( QStringLiteral( "Transformation of extent failed!" ) );
+  }
+
+  bool canceled = false;
+  if ( mSubIndexes.isEmpty() )
+  {
+    canceled = !renderIndex( pc );
+  }
+  else
+  {
+    for ( const auto &si : mSubIndexes )
+    {
+      if ( canceled )
+        break;
+
+      QgsPointCloudIndex *pc = si.index.get();
+
+      if ( !renderExtent.intersects( si.extent ) )
+        continue;
+
+      if ( !pc || !pc->isValid() || renderExtent.width() > si.extent.width() )
+      {
+        // when dealing with virtual point clouds, we want to render the individual extents when zoomed out
+        // and only use the selected renderer when zoomed in
+        mSubExtentsRenderer->startRender( context );
+        mSubExtentsRenderer->renderExtent( si.geometry, context );
+        mSubExtentsRenderer->stopRender( context );
```

Comment:
the startRender()/stopRender() calls should go before/after the for cycle...

Reply:
ok, there will be cases where the renderer starts and stops without rendering anything, but I guess that's no issue